### PR TITLE
struct serial_ops: make functions optional

### PR DESCRIPTION
**Pull request by jforissier**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
Note: force pushed to restart CI on an updated merge commit with master. It looks like restarting the CI via the GUI ("Re-run failed jobs") won't cause the merge commit to change although master has moved since.

**Review by xiaoxuZeng — APPROVED** [review]

File: core/arch/arm/plat-vexpress/main.c [hunk 1]
```diff
@@ -101,7 +101,7 @@ static void read_console(void)
 {
 	struct serial_chip *cons = &console_data.chip;
 
-	while (cons->ops->have_rx_data(cons)) {
+	while (cons->ops->have_rx_data && cons->ops->have_rx_data(cons)) {
```
Comment (etienne-lms) [inline core/arch/arm/plat-vexpress/main.c:104]: check also cons->ops->getchar is not NULL.
maybe:
```patch
+	if (!cons->ops->getchar || !cons->ops->have_rx_data)
+		return;
+
 	while (cons->ops->have_rx_data(cons)) {
 		int ch __maybe_unused = cons->ops->getchar(cons);
 		(...)
```

File: core/include/drivers/serial.h [hunk 2]
```diff
@@ -16,7 +16,9 @@ struct serial_chip {
 };
 
 struct serial_ops {
+	/* Must not be NULL */
```
Comment (etienne-lms) [inline core/include/drivers/serial.h:19]: suggestion:
here:	`/* Mandatory handlers */`
and below: 	`/* Optional handlers */`

**Comment by jforissier** [comment]
Updated

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Pull request merged**